Fix async engine reloading

Steveplays28 opened the pull request:
Also made the `/ambient-reload` command reload the engine asynchronously.
Follow-up PR to #246, should avoid any resource reload (compatibility) issues in the future such as https://github.com/IMB11/Fog/issues/23.

If you spam the command/resource reload fast enough it'll keep reloading the engine with this code, though I'm not sure if that's a big issue requiring a fix currently. You'd have to check if a `CompletableFuture<Void>` field is already reloading the engine.

Steveplays28 committed:
fix: Fix async engine reloading

Also made the `/ambient-reload` command reload the engine asynchronously.

CreativeMD commented:
Thanks for your work! I'm not sure however if creating a thread pool for this rare task is not unnecessary overhead. I think it would be enough to execute this on the Util.backgroundExecutor() for example.

Steveplays28 commented:
@CreativeMD I didn't consider that, that's probably fair, but the reloading needs to happen on 1 thread, because I don't think it's thread-safe code per-se.
If you make it execute on the resource reloader's thread it has the chance to execute thread-unsafely across multiple threads, because it takes a random background worker thread, and if you make it execute on the main thread (apply stage) it blocks the game's rendering/ticking.

Does `Util.backgroundExecutor()` use 1 thread or all the background worker threads?

CreativeMD commented:
You could synchronize the load of the engine. I think adding synchronized to AmbientEngine.loadAmbientSoundEngine(String) should be enough.

Steveplays28 committed:
refactor: Synchronize `AmbientSounds#reload` and `AmbientEngine#loadAmbientEngine`

Also replaced the Ambient Sounds engine reload thread pool executor with `Util#backgroundExecutor`.

Steveplays28 commented:
@CreativeMD I've synchronized `AmbientSounds#reload` and `AmbientEngine#loadAmbientEngine`, because both of those aren't thread-safe. I've also replaced the custom thread pool executor with `Util#backgroundExecutor`.

Steveplays28 commented:
Thanks for merging and the feedback CreativeMD!
